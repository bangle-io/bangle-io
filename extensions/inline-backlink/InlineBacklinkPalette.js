import React, { useMemo } from 'react';
import reactDOM from 'react-dom';
import {
  useInlinePaletteItems,
  useInlinePaletteQuery,
} from 'inline-palette/index';
import { SidebarRow } from 'ui-components';
import { replaceSuggestionMarkWith } from 'inline-palette/inline-palette';
import { resolvePath, useGetCachedWorkspaceFiles } from 'workspace/index';
import { backLinkNodeName, palettePluginKey } from './config';

export function InlineBacklinkPalette() {
  const { query, counter } = useInlinePaletteQuery(palettePluginKey);
  const [currentFiles = []] = useGetCachedWorkspaceFiles();

  const items = useMemo(() => {
    if (!query) {
      return [];
    }

    const items = currentFiles.map((wsPath) => {
      return {
        uid: wsPath,
        title: wsPath,
        editorExecuteCommand: ({ item }) => {
          return (state, dispatch, view) => {
            let nodeType = state.schema.nodes[backLinkNodeName];
            return replaceSuggestionMarkWith(
              palettePluginKey,
              nodeType.create({
                title: resolvePath(wsPath).fileName,
                wsPath: wsPath,
              }),
            )(state, dispatch, view);
          };
        },
      };
    });
    return items.filter((item) => queryMatch(item, query)).slice(0, 50);
  }, [query, currentFiles]);

  const { tooltipContentDOM, getItemProps } = useInlinePaletteItems(
    palettePluginKey,
    items,
    counter,
  );

  return reactDOM.createPortal(
    <div className="bangle-emoji-suggest">
      {query === '' && (
        <SidebarRow
          dataId="searchNote"
          className="palette-row"
          title="Search for a note"
        />
      )}
      {items.map((item, i) => {
        return (
          <SidebarRow
            key={item.uid}
            dataId={item.uid}
            className="palette-row"
            title={item.title}
            rightHoverIcon={item.rightHoverIcon}
            rightIcon={
              <kbd className="whitespace-nowrap">{item.keybinding}</kbd>
            }
            {...getItemProps(item, i)}
          />
        );
      })}
    </div>,
    tooltipContentDOM,
  );
}

function queryMatch(item, query) {
  const keywords = item.keywords || '';

  if (keywords.length > 0) {
    if (strMatch(keywords.split(','), query)) {
      return item;
    }
  }
  return strMatch(item.title, query) ? item : undefined;
}

function strMatch(a, b) {
  b = b.toLocaleLowerCase();
  if (Array.isArray(a)) {
    return a.filter(Boolean).some((str) => strMatch(str, b));
  }

  a = a.toLocaleLowerCase();
  return a.includes(b) || b.includes(a);
}
