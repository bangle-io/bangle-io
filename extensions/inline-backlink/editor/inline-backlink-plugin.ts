import { NodeView } from '@bangle.dev/core';

import { backlinkNodeName } from '../config';

export function inlineBacklinkPlugin() {
  return () => [
    NodeView.createPlugin({
      name: backlinkNodeName,
      // inline-block allows the span to get full height of image
      // or else folks depending on the boundingBox get incorrect
      // dimensions.
      containerDOM: ['span', { style: 'display: inline-block;' }],
    }),
  ];
}
