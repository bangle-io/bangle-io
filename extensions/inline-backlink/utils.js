export function removeMdExtension(str) {
  if (str.endsWith('.md')) {
    return str.slice(0, -3);
  }
  return str;
}
