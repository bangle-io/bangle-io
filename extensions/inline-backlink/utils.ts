import {
  filePathToWsPath,
  removeExtension,
  resolvePath,
  sanitizeFilePath,
  suffixWithNoteExtension,
  validateWsPath,
} from '@bangle.io/ws-path';

export function wsPathFromQuery(query: string, wsName: string) {
  let filePath = query.split(']]').join('');
  filePath = filePath.trim();
  filePath = suffixWithNoteExtension(filePath);
  let wsPath = filePathToWsPath(wsName, filePath);
  try {
    validateWsPath(wsPath);
  } catch (error) {
    wsPath = filePathToWsPath(wsName, sanitizeFilePath(filePath));
  }

  return wsPath;
}

export function getBacklinkPath(wsPath: string, allWsPaths: readonly string[]) {
  const { fileName, filePath } = resolvePath(wsPath);
  const matchingFilenames = allWsPaths.filter(
    (w) => resolvePath(w).fileName === fileName,
  );

  if (matchingFilenames.length === 0) {
    return removeExtension(filePath);
  }

  // if there are multiple files with the same name
  // give it an absolute path
  if (matchingFilenames.length > 1) {
    return removeExtension(filePath);
  }

  return removeExtension(fileName);
}
