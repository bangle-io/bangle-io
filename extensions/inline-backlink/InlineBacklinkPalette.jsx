import React, { useMemo, useEffect, useContext } from 'react';
import reactDOM from 'react-dom';
import { useEditorViewContext } from '@bangle.dev/react';
import {
  useInlinePaletteItems,
  useInlinePaletteQuery,
  replaceSuggestionMarkWith,
} from 'inline-palette/index';
import { UniversalPalette, InlinePaletteRow } from 'ui-components/index';
import {
  resolvePath,
  filePathToWsPath,
  validateWsPath,
  sanitizeFilePath,
} from 'ws-path';
import { backLinkNodeName, palettePluginKey } from './config';
import { conditionalSuffix, removeMdExtension } from 'utils/index';
import { useWorkspaceContext } from 'workspace-context/index';

// Creating this also closes the palette
const createBackLinkNode = (wsPath, allNoteWsPaths) => {
  return (state, dispatch, view) => {
    const nodeType = state.schema.nodes[backLinkNodeName];
    const backLinkPath = getBacklinkPath(wsPath, allNoteWsPaths);

    return replaceSuggestionMarkWith(
      palettePluginKey,
      nodeType.create({
        path: backLinkPath,
      }),
    )(state, dispatch, view);
  };
};
export function InlineBacklinkPalette() {
  const { query, counter, tooltipContentDOM } =
    useInlinePaletteQuery(palettePluginKey);

  return reactDOM.createPortal(
    <div className="inline-palette-wrapper shadow-2xl">
      <div className="inline-palette-items-wrapper">
        {/* TODO I am unable to hide inner when palette is invisible */}
        <InlineBacklinkPaletteInner query={query} counter={counter} />
        <UniversalPalette.PaletteInfo>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">â†‘â†“</kbd> Navigate
          </UniversalPalette.PaletteInfoItem>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">Enter</kbd> Create link
          </UniversalPalette.PaletteInfoItem>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">Esc</kbd> Dismiss
          </UniversalPalette.PaletteInfoItem>
        </UniversalPalette.PaletteInfo>
      </div>
    </div>,
    tooltipContentDOM,
  );
}

function InlineBacklinkPaletteInner({ query, counter }) {
  const view = useEditorViewContext();
  const { wsName, noteWsPaths = [] } = useWorkspaceContext();
  const items = useMemo(() => {
    return filterItems(wsName, query, noteWsPaths);
  }, [query, noteWsPaths, wsName]);

  const { getItemProps } = useInlinePaletteItems(
    palettePluginKey,
    items,
    counter,
    () => false,
  );

  useEffect(() => {
    if (query === ']]' || query.includes('[[')) {
      replaceSuggestionMarkWith(palettePluginKey, '')(
        view.state,
        view.dispatch,
        view,
      );
    } else if (query.endsWith(']]')) {
      createBackLinkNode(wsPathFromQuery(query, wsName), noteWsPaths)(
        view.state,
        view.dispatch,
        view,
      );
    }
  }, [query, wsName, view, noteWsPaths]);

  return (
    <>
      {query === '' && (
        <InlinePaletteRow
          dataId="searchNote"
          className="palette-row"
          title={'ðŸ’¡ Search for a note to link it'}
        />
      )}
      {items.map((item, i) => {
        return (
          <InlinePaletteRow
            key={item.uid}
            dataId={item.uid}
            className="palette-row"
            description={item.title}
            rightHoverIcon={item.rightHoverIcon}
            rightIcon={
              <kbd className="whitespace-nowrap">{item.keybinding}</kbd>
            }
            {...getItemProps(item, i)}
          />
        );
      })}
    </>
  );
}

export function filterItems(wsName, query, allNoteWsPaths) {
  if (!query) {
    return [];
  }

  const items = allNoteWsPaths.map((wsPath) => {
    return {
      wsPath,
      uid: wsPath,
      title: removeMdExtension(resolvePath(wsPath).filePath),
      editorExecuteCommand: () => {
        return createBackLinkNode(wsPath, allNoteWsPaths);
      },
    };
  });

  let filteredItems = items
    .filter((item) => queryMatch(item, query))
    .slice(0, 15);

  const exactMatch = filteredItems.find(
    (item) =>
      item.title === query ||
      conditionalSuffix(query, '.md') === resolvePath(item.wsPath).filePath,
  );
  if (!exactMatch && query.length > 0) {
    const wsPath = wsPathFromQuery(query, wsName);

    filteredItems.unshift({
      uid: 'create-' + wsPath,
      title: 'Create: ' + removeMdExtension(resolvePath(wsPath).filePath),
      editorExecuteCommand: () => {
        return createBackLinkNode(wsPath, allNoteWsPaths);
      },
    });
  }

  return filteredItems;
}

function queryMatch(item, query) {
  const keywords = item.keywords || '';

  if (keywords.length > 0) {
    if (strMatch(keywords.split(','), query)) {
      return item;
    }
  }
  return strMatch(item.title, query) ? item : undefined;
}

function strMatch(a, b) {
  b = b.toLocaleLowerCase();
  if (Array.isArray(a)) {
    return a.filter(Boolean).some((str) => strMatch(str, b));
  }

  a = a.toLocaleLowerCase();
  return a.includes(b) || b.includes(a);
}

export function getBacklinkPath(wsPath, allWsPaths) {
  const { fileName, filePath } = resolvePath(wsPath);
  const matchingFilenames = allWsPaths.filter(
    (w) => resolvePath(w).fileName === fileName,
  );

  if (matchingFilenames.length === 0) {
    return removeMdExtension(filePath);
  }

  // if there are multiple files with the same name
  // give it an absolute path
  if (matchingFilenames.length > 1) {
    return removeMdExtension(filePath);
  }

  return removeMdExtension(fileName);
}

function wsPathFromQuery(query, wsName) {
  let filePath = query.split(']]').join('');
  filePath = filePath.trim();
  filePath = conditionalSuffix(filePath, '.md');
  let wsPath = filePathToWsPath(wsName, filePath);
  try {
    validateWsPath(wsPath);
  } catch (error) {
    wsPath = filePathToWsPath(wsName, sanitizeFilePath(filePath));
  }

  return wsPath;
}
