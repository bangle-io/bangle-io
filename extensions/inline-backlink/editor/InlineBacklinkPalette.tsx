import React, { useEffect, useMemo } from 'react';
import reactDOM from 'react-dom';

import type { EditorState, EditorView } from '@bangle.dev/pm';
import { useEditorViewContext } from '@bangle.dev/react';

import { nsmApi2 } from '@bangle.io/api';
import { byLengthAsc, useFzfSearch } from '@bangle.io/fzf-search';
import {
  replaceSuggestionMarkWith,
  useInlinePaletteItems,
  useInlinePaletteQuery,
} from '@bangle.io/inline-palette';
import { InlinePaletteRow, UniversalPalette } from '@bangle.io/ui-components';
import { assertNotUndefined, insertAt } from '@bangle.io/utils';
import {
  removeExtension,
  resolvePath,
  suffixWithNoteExtension,
} from '@bangle.io/ws-path';

import { backlinkNodeName, palettePluginKey } from '../config';
import { getBacklinkPath, wsPathFromQuery } from '../utils';

const FZF_SEARCH_LIMIT = 12;

// Creating this also closes the palette
const createBacklinkNode = (
  wsPath: string,
  allNoteWsPaths: readonly string[],
) => {
  return (
    state: EditorState,
    dispatch: EditorView['dispatch'] | undefined,
    view: EditorView | undefined,
  ) => {
    const nodeType = state.schema.nodes[backlinkNodeName];
    const backlinkPath = getBacklinkPath(wsPath, allNoteWsPaths);

    assertNotUndefined(nodeType, 'wikiLink must be defined');

    return replaceSuggestionMarkWith(
      palettePluginKey,
      nodeType.create({
        path: backlinkPath,
      }),
    )(state, dispatch, view);
  };
};

export function InlineBacklinkPalette() {
  const view = useEditorViewContext();

  if (!view || view.isDestroyed) {
    return null;
  }

  return <InlineBacklinkPalettePortal />;
}

function InlineBacklinkPalettePortal() {
  const { query, counter, tooltipContentDOM, isVisible } =
    useInlinePaletteQuery(palettePluginKey);

  return reactDOM.createPortal(
    <InlineBacklinkPaletteWrapper
      query={query}
      counter={counter}
      isVisible={isVisible}
      editorView={useEditorViewContext()}
    />,
    tooltipContentDOM,
  );
}

export function InlineBacklinkPaletteWrapper({
  query,
  counter,
  isVisible,
  editorView,
}: {
  query: string;
  counter: number;
  isVisible: boolean;
  editorView: EditorView;
}) {
  return (
    <div className="shadow-2xl B-ui-components_inline-palette-wrapper flex flex-col bg-colorNeutralBgLayerFloat">
      <div className="B-ui-components_inline-palette-items-wrapper">
        {/* TODO I am unable to hide inner when palette is invisible */}
        {isVisible && (
          <InlineBacklinkPaletteInner
            query={query}
            counter={counter}
            editorView={editorView}
          />
        )}
        <UniversalPalette.PaletteInfo>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">â†‘â†“</kbd> Navigate
          </UniversalPalette.PaletteInfoItem>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">Enter</kbd> Create link
          </UniversalPalette.PaletteInfoItem>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">Esc</kbd> Dismiss
          </UniversalPalette.PaletteInfoItem>
        </UniversalPalette.PaletteInfo>
      </div>
    </div>
  );
}

const EMPTY_ARRAY: string[] = [];

function InlineBacklinkPaletteInner({
  query,
  editorView: view,
  counter,
}: {
  query: string;
  counter: number;
  editorView: EditorView;
}) {
  const { wsName, noteWsPaths = EMPTY_ARRAY } =
    nsmApi2.workspace.useWorkspace();

  const match = useFzfSearch(noteWsPaths, query, {
    limit: FZF_SEARCH_LIMIT,
    selector: (item: any) => resolvePath(item, true).filePath,
    tiebreakers: [byLengthAsc],
  });

  const items = useMemo(() => {
    let res = match.map((r) => {
      const wsPath = r.item;

      return {
        wsPath: wsPath,
        uid: wsPath,
        title: removeExtension(resolvePath(wsPath, true).filePath),
        editorExecuteCommand: () => {
          return createBacklinkNode(wsPath, noteWsPaths);
        },
      };
    });

    const exactMatch = res.find(
      (item) =>
        item.title === query ||
        suffixWithNoteExtension(query) ===
          resolvePath(item.wsPath, true).filePath,
    );

    if (!exactMatch && query.length > 0 && wsName) {
      const wsPath = wsPathFromQuery(query, wsName);

      let createItem = {
        uid: 'create-' + wsPath,
        wsPath: wsPath,
        title: 'Create: ' + removeExtension(resolvePath(wsPath).filePath),
        editorExecuteCommand: () => {
          return createBacklinkNode(wsPath, noteWsPaths);
        },
      };
      res = insertAt(res, 1, createItem);
    }

    return res;
  }, [noteWsPaths, match, wsName, query]);

  const { getItemProps } = useInlinePaletteItems(
    palettePluginKey,
    items,
    counter,
    () => false,
  );

  useEffect(() => {
    if (!wsName) {
      return;
    }
    if (query === ']]' || query.includes('[[')) {
      replaceSuggestionMarkWith(palettePluginKey, '')(
        view.state,
        view.dispatch,
        view,
      );
    } else if (query.endsWith(']]')) {
      createBacklinkNode(wsPathFromQuery(query, wsName), noteWsPaths)(
        view.state,
        view.dispatch,
        view,
      );
    }
  }, [query, wsName, view, noteWsPaths]);

  return (
    <>
      {query === '' && (
        <InlinePaletteRow
          dataId="searchNote"
          className="palette-row"
          title={'ðŸ’¡ Search for a note to link it'}
        />
      )}
      {items.map((item, i) => {
        return (
          <InlinePaletteRow
            key={item.uid}
            dataId={item.uid}
            className="palette-row"
            description={item.title}
            {...getItemProps(item, i)}
          />
        );
      })}
    </>
  );
}
