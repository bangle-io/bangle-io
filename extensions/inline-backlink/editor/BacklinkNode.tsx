import React, { useEffect, useState } from 'react';

import type { EditorState } from '@bangle.dev/pm';

import { nsmApi2 } from '@bangle.io/api';
import { EditorDisplayType } from '@bangle.io/constants';
import { getEditorPluginMetadata } from '@bangle.io/editor-common';
import type { RenderReactNodeView } from '@bangle.io/extension-registry';
import type { WsPath } from '@bangle.io/shared-types';
import { useHover, useTooltipPositioner } from '@bangle.io/ui-components';

import { BacklinkNodeButton } from '../BacklinkNodeButton';
import { calcWikiLinkMapping, getAllWikiLinks } from '../calculate-matches';
import { backlinkNodeName } from '../config';
import { LinkPreview } from './LinkPreview';
import { useOnClickBacklink } from './use-on-click-backlink';

export function BacklinkNode({
  nodeAttrs,
  editorState,
}: {
  nodeAttrs: { path: string; title?: string };
  editorState: EditorState;
}) {
  const { wsPath: currentWsPath, editorDisplayType } =
    getEditorPluginMetadata(editorState);

  const { wsName, noteWsPaths } = nsmApi2.workspace.useWorkspace();

  let { path: wikiLink, title } = nodeAttrs;
  title = title || wikiLink;

  const onClick = useOnClickBacklink({
    wikiLink,
    currentWsPath,
    editorState,
  });

  const [backlinksWsPath, updateBacklinksWsPath] = useState<WsPath | undefined>(
    undefined,
  );

  const disablePopup = editorDisplayType === EditorDisplayType.Floating;

  const [isHovered, updateIsHovered] = useState(false);
  const { hoverProps: tooltipHoverProps, isHovered: isTooltipHovered } =
    useHover({ isDisabled: disablePopup });

  const {
    isTooltipVisible,
    setTooltipElement,
    setTriggerElement,
    tooltipProps,
  } = useTooltipPositioner({
    isDisabled: disablePopup,
    isActive: !disablePopup && (isHovered || isTooltipHovered),
    xOffset: 10,
    yOffset: 0,
    // TODO we can optimize where we position based on where empty
    // space exists
    placement: 'right',
    delay: 450,
    immediateClose: false,
  });

  useEffect(() => {
    if (noteWsPaths) {
      const result = calcWikiLinkMapping(
        noteWsPaths,
        getAllWikiLinks(editorState),
      ).get(wikiLink);

      updateBacklinksWsPath(result);
    }
  }, [wsName, noteWsPaths, wikiLink, editorState]);

  useEffect(() => {
    let targetWsPath = backlinksWsPath;
    let didSet = false;

    if (targetWsPath && isTooltipVisible) {
      didSet = true;
      nsmApi2.workspace.pushOpenedWsPath((openedWsPaths) => {
        return openedWsPaths.updatePopupEditorWsPath(targetWsPath);
      });
    }

    return () => {
      if (didSet) {
        nsmApi2.workspace.pushOpenedWsPath((openedWsPaths) => {
          // make sure we are unsetting the correct wsPath
          if (
            targetWsPath &&
            targetWsPath === openedWsPaths.popupEditorWsPath
          ) {
            return openedWsPaths.updatePopupEditorWsPath(undefined);
          }

          return openedWsPaths;
        });
      }
    };
  }, [backlinksWsPath, isTooltipVisible]);

  return (
    <>
      <BacklinkNodeButton
        ref={setTriggerElement}
        linkNotFound={!backlinksWsPath}
        title={title}
        onClick={onClick}
        onHoverChange={updateIsHovered}
      />
      {isTooltipVisible && backlinksWsPath && (
        <LinkPreview
          disablePreview={disablePopup}
          positionProps={{ ...tooltipHoverProps, ...tooltipProps.attributes }}
          ref={setTooltipElement}
          style={tooltipProps.style}
          wsPath={backlinksWsPath}
        />
      )}
    </>
  );
}

export const renderReactNodeView: RenderReactNodeView = {
  [backlinkNodeName]: ({ nodeViewRenderArg }) => {
    const { path, title } = nodeViewRenderArg.node.attrs;

    if (typeof path !== 'string') {
      return <span>Invalid Path</span>;
    }

    return (
      <BacklinkNode
        nodeAttrs={{ path, title }}
        editorState={nodeViewRenderArg.view.state}
      />
    );
  },
};
