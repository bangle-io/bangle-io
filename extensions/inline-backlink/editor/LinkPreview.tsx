import React from 'react';

import { internalApi, vars } from '@bangle.io/api';
import { PopupEditor } from '@bangle.io/editor';
import type { WsPath } from '@bangle.io/shared-types';

export const LinkPreview = React.forwardRef<
  HTMLDivElement,
  {
    disablePreview: boolean;
    positionProps: Parameters<
      typeof PopupEditor
    >[0]['popupContainerProps']['positionProps'];
    style?: React.CSSProperties;
    wsPath: WsPath;
  }
>(({ positionProps, wsPath, style }, ref) => {
  return (
    <PopupEditor
      ref={ref}
      editorProps={{
        eternalVars: internalApi.eternalVars.getEternalVars(),
        wsPath: wsPath,
      }}
      popupContainerProps={{
        style: {
          // setting width and height are important otherwise
          // tippy confuses and causes issue with scroll
          ...style,
          width: vars.misc.miniEditorWidth,
          height: 400,
        },
        className:
          'B-inline-backlink_popup-editor z-popup border-1 border-solid border-colorNeutralBorder shadow-2xl',
        positionProps: positionProps,
      }}
    />
  );
});
