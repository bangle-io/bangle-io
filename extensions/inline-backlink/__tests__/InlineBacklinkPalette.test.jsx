import { defaultPlugins, defaultSpecs } from '@bangle.dev/all-base-components';
import { Node, PluginKey } from '@bangle.dev/pm';
import { useEditorViewContext } from '@bangle.dev/react';
import { act, render } from '@testing-library/react';
import { Extension, ExtensionRegistry } from 'extension-registry';
import {
  replaceSuggestionMarkWith,
  useInlinePaletteItems,
  useInlinePaletteQuery,
} from 'inline-palette';
import React from 'react';
import { sleep } from 'utils/utility';
import { useWorkspaceContext } from 'workspace-context';
import inlineBackLinkExtension from '../index';
import { InlineBacklinkPalette } from '../editor/InlineBacklinkPalette';

jest.mock('@bangle.dev/react', () => {
  return {
    useEditorViewContext: jest.fn(() => jest.fn()),
  };
});

jest.mock('inline-palette', () => {
  const otherThings = jest.requireActual('inline-palette');

  return {
    ...otherThings,
    useInlinePaletteItems: jest.fn(),
    useInlinePaletteQuery: jest.fn(),
    replaceSuggestionMarkWith: jest.fn(),
  };
});

jest.mock('workspace-context/index', () => {
  return {
    ...jest.requireActual('workspace-context/index'),
    useWorkspaceContext: jest.fn(),
  };
});

const coreExtension = Extension.create({
  name: 'core',
  editor: {
    specs: defaultSpecs(),
    plugins: defaultPlugins(),
  },
});

const extensionRegistry = new ExtensionRegistry([
  coreExtension,
  inlineBackLinkExtension,
]);

const schema = extensionRegistry.specRegistry.schema;
const mockView = {
  state: {
    schema,
  },
};
let tooltipContentDOM, query;
beforeEach(async () => {
  tooltipContentDOM = document.createElement('div');
  query = '';
  replaceSuggestionMarkWith.mockImplementation(() => jest.fn());
  useEditorViewContext.mockImplementation(() => {
    return mockView;
  });
  useWorkspaceContext.mockImplementation(() => ({
    wsName: 'test-ws',
    noteWsPaths: [],
  }));
  useInlinePaletteItems.mockImplementation(() => {
    return {
      getItemProps: jest.fn(),
    };
  });
  useInlinePaletteQuery.mockImplementation(() => ({
    tooltipContentDOM,
    query,
    counter: 0,
    isVisible: true,
  }));
});

test('Initial render', async () => {
  const noteWsPaths = ['test-ws:my-file.md'];

  useWorkspaceContext.mockImplementation(() => ({
    noteWsPaths,
  }));

  const result = render(<InlineBacklinkPalette />);

  expect(tooltipContentDOM).toMatchSnapshot();
});

test('Hides when isVisible is false', async () => {
  useInlinePaletteQuery.mockImplementation(() => ({
    tooltipContentDOM,
    query,
    counter: 0,
    isVisible: false,
  }));

  const noteWsPaths = ['test-ws:my-file.md'];

  useWorkspaceContext.mockImplementation(() => ({
    noteWsPaths,
  }));

  const result = render(<InlineBacklinkPalette />);

  expect(tooltipContentDOM).toMatchSnapshot();
});

test('Renders palette rows correctly', async () => {
  const noteWsPaths = ['test-ws:my-file.md'];

  useWorkspaceContext.mockImplementation(() => ({
    noteWsPaths,
  }));
  query = 'my';

  const result = render(<InlineBacklinkPalette />);

  let rows = Array.from(tooltipContentDOM.querySelectorAll('.palette-row'));

  expect(
    rows.find((node) => node.textContent.includes('my-file')),
  ).toBeTruthy();

  // doesnt show a row when query doesn't match
  query = 'no';
  result.rerender(<InlineBacklinkPalette />);
  rows = Array.from(tooltipContentDOM.querySelectorAll('.palette-row'));
  expect(rows.find((node) => node.textContent.includes('my-file'))).toBe(
    undefined,
  );
});

test('Renders filtering correct', async () => {
  const noteWsPaths = ['test-ws:my-file.md', 'test-ws:my-other-file.md'];

  useWorkspaceContext.mockImplementation(() => ({
    noteWsPaths,
  }));
  query = 'my';

  const result = render(<InlineBacklinkPalette />);

  expect(
    Array.from(tooltipContentDOM.querySelectorAll('.palette-row')).map(
      (node) => node.textContent,
    ),
  ).toEqual(['my-file', 'Create: my', 'my-other-file']);

  query = 'other';
  result.rerender(<InlineBacklinkPalette />);
  expect(
    Array.from(tooltipContentDOM.querySelectorAll('.palette-row')).map(
      (node) => node.textContent,
    ),
  ).toEqual(['my-other-file', 'Create: other']);
});

test('If exact match does not show create', async () => {
  const noteWsPaths = ['test-ws:my-file.md', 'test-ws:my-other-file.md'];

  useWorkspaceContext.mockImplementation(() => ({
    noteWsPaths,
  }));
  query = 'my-file';

  render(<InlineBacklinkPalette />);

  expect(
    Array.from(tooltipContentDOM.querySelectorAll('.palette-row')).map(
      (node) => node.textContent,
    ),
  ).toEqual(['my-file', 'my-other-file']);
});

test('Handles malformed query', async () => {
  const noteWsPaths = ['test-ws:my-file.md'];

  useWorkspaceContext.mockImplementation(() => ({
    noteWsPaths,
  }));
  query = ']]';

  const result = render(<InlineBacklinkPalette />);
  const prom = sleep();
  await act(() => prom);
  expect(replaceSuggestionMarkWith).toBeCalledTimes(1);
  expect(replaceSuggestionMarkWith).nthCalledWith(1, expect.any(PluginKey), '');
});

test('Creates a backlink node when closed by typing ]]', async () => {
  const noteWsPaths = ['test-ws:my-file.md'];
  useWorkspaceContext.mockImplementation(() => ({
    noteWsPaths,
  }));
  // NOTE: its not [[better]], because [[ is part of the suggest query mark
  query = 'better]]';
  render(<InlineBacklinkPalette />);
  const prom = sleep();
  await act(() => prom);
  expect(replaceSuggestionMarkWith).toBeCalledTimes(1);
  expect(replaceSuggestionMarkWith).nthCalledWith(
    1,
    expect.any(PluginKey),
    expect.any(Node),
  );

  expect(replaceSuggestionMarkWith.mock.calls[0][1].attrs).toEqual({
    path: 'better',
    title: null,
  });
});
