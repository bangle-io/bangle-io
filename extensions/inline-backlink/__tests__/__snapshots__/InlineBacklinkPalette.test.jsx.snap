// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hides when isVisible is false 1`] = `
<div>
  <div
    class="inline-palette-wrapper shadow-2xl"
  >
    <div
      class="inline-palette-items-wrapper"
    >
      <div
        class="flex flex-row justify-center my-1 space"
      >
        <span
          class="text-xs ml-2 mr-2 font-light"
        >
          <kbd
            class="font-normal"
          >
            â†‘â†“
          </kbd>
           Navigate
        </span>
        <span
          class="text-xs ml-2 mr-2 font-light"
        >
          <kbd
            class="font-normal"
          >
            Enter
          </kbd>
           Create link
        </span>
        <span
          class="text-xs ml-2 mr-2 font-light"
        >
          <kbd
            class="font-normal"
          >
            Esc
          </kbd>
           Dismiss
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Initial render 1`] = `
<div>
  <div
    class="inline-palette-wrapper shadow-2xl"
  >
    <div
      class="inline-palette-items-wrapper"
    >
      <div
        class="flex flex-row items-center inline-palette-row cursor-pointer palette-row hover-allowed"
        data-id="searchNote"
        style="padding-left: 16px; padding-right: 16px;"
      >
        <span
          class="flex flex-col"
        >
          <span
            class="text-base font-bold truncate select-none"
          >
            ðŸ’¡ Search for a note to link it
          </span>
          <span
            class="text-sm font-normal select-none"
          />
        </span>
        <span
          class=" flex flex-1"
        />
      </div>
      <div
        class="flex flex-row items-center inline-palette-row cursor-pointer palette-row hover-allowed"
        data-id="test-ws:my-file.md"
        style="padding-left: 16px; padding-right: 16px;"
      >
        <span
          class="flex flex-col"
        >
          <span
            class="text-base font-bold truncate select-none"
          />
          <span
            class="text-sm font-normal select-none"
          >
            my-file
          </span>
        </span>
        <span
          class=" flex flex-1"
        />
      </div>
      <div
        class="flex flex-row justify-center my-1 space"
      >
        <span
          class="text-xs ml-2 mr-2 font-light"
        >
          <kbd
            class="font-normal"
          >
            â†‘â†“
          </kbd>
           Navigate
        </span>
        <span
          class="text-xs ml-2 mr-2 font-light"
        >
          <kbd
            class="font-normal"
          >
            Enter
          </kbd>
           Create link
        </span>
        <span
          class="text-xs ml-2 mr-2 font-light"
        >
          <kbd
            class="font-normal"
          >
            Esc
          </kbd>
           Dismiss
        </span>
      </div>
    </div>
  </div>
</div>
`;
